src/main/java/com/indeed/jiraactions/api/ApiCaller.java

```diff
@@ -22,159 +28,102 @@
                     }
                     sb.append("\"").append(key).append("\": \"").append(value).append("\",");
                 }
-            }
-            sb.append("}");
-
-            sb.append("}");
+                sb.append('}');
+                sb.append('}');
 
-            if (urlConnection != null) {
                 sb.append(", \"Response\": {");
-                for (final Map.Entry<String, List<String>> header : urlConnection.getHeaderFields().entrySet()) {
-                    final String key = header.getKey();
+                for (int i = 0, size = responseHeaders.size(); i < size; i++) {
+                    final String key = responseHeaders.name(i);
                     final String value;
                     if ("Set-Cookie".equals(key)) {
                         value = "<Omitted>";
                     } else {
-                        value = String.join(",", header.getValue());
+                        value = String.join(",", responseHeaders.value(i));
                     }
                     sb.append("\"").append(key).append("\": \"").append(value).append("\",");
                 }
-                sb.append("\"Code\": ").append(urlConnection.getResponseCode()).append(",");
-                sb.append("\"Message\": \"").append(urlConnection.getResponseMessage()).append("\",");
-            }
-            if (apiResults != null) {
-                sb.append("\"Body\": \"").append(apiResults).append("\"");
-            }
-            if (urlConnection != null) {
-                final InputStream error = urlConnection.getErrorStream();
-                if(error != null) {
-                    final BufferedReader errorBr = new BufferedReader(new InputStreamReader(error));
-                    sb.append("\"Error Body\": \"");
-                    String line = errorBr.readLine();
-                    while (line != null) {
-                        sb.append(line).append(System.lineSeparator());
-                        line = errorBr.readLine();
-                    }
-                    errorBr.close();
-                } else {
-                    sb.append("Unable to open error stream for reading");
-                }
-                sb.append("\"");
-            }
-
-            sb.append("}");
-            sb.append("}");
-            log.error("Encountered connection error: " + sb);
-            throw e;
-        } finally {
-            if (br != null) {
-                try {
-                    br.close();
-                } catch (final IOException ignored) {
-                }
+                sb.append("\"Code\": ").append(response.code()).append(",");
+                sb.append("\"Message\": \"").append(response.message()).append("\",");
+                sb.append("\"Error Body\": \"").append(response.body().string()).append("\"");
+                sb.append('}');
+                sb.append('}');
+                log.error("Encountered connection error: " + sb);
```

Comment:
So let's say an error occurs. We log the error, then what happens?

Reply:
It logs the request attempt and then throws the IOException

Reply:
If the HTTP request completes without throwing an exception but `response.isSuccessful()` returns false (meaning it has a 4xx or 5xx response code), your code will log the error and then attempt to deserialize the responseBody and return that. I suspect that will cause an exception, but I'm not positive.

If the HTTP request completes exceptionally (because of connectivity problems or timeouts), your code will not log anything here at all and rely on the caller catching the exception.


There's some nuance in the old connection code you may be missing; `HttpURLConnection.getInputStream()` would throw an `IOException` if the server you were talking to returned a 4xx or a 5xx code.